src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UIMetadataImportService.java

```diff
@@ -1,172 +1,242 @@
+	public Map<String, Object> setProjectSampleMetadataSampleId(HttpSession session, long projectId,
+			String sampleNameColumn) {
+		// Attempt to get the metadata from the sessions
+		SampleMetadataStorage stored = (SampleMetadataStorage) session.getAttribute("pm-" + projectId);
+
+		if (stored != null) {
+			stored.setSampleNameColumn(sampleNameColumn);
+			Project project = projectService.read(projectId);
+			List<Map<String, String>> rows = stored.getRows();
+
+			// Remove 'rows' since they are now going to be sorted into found and not found.
+			stored.removeRows();
+			List<Map<String, String>> found = new ArrayList<>();
+			List<Map<String, String>> missing = new ArrayList<>();
+
+			// Get the metadata out of the table.
+			for (Map<String, String> row : rows) {
+				try {
+					// If this throws an error than the sample does not exist.
+					sampleService.getSampleBySampleName(project, row.get(sampleNameColumn));
+					found.add(row);
+				} catch (EntityNotFoundException e) {
+					missing.add(row);
+				}
+			}
+
+			stored.saveFound(found);
+			stored.saveMissing(missing);
 		}
 
-		// Only look at the first sheet in the workbook as this should be the file we want.
-		Sheet sheet = workbook.getSheetAt(0);
-		Iterator<Row> rowIterator = sheet.iterator();
-
-		List<String> headers = getWorkbookHeaders(rowIterator.next());
-		storage.saveHeaders(headers);
-
-		// Get the metadata out of the table.
-		List<Map<String, String>> rows = new ArrayList<>();
-		while (rowIterator.hasNext()) {
-			Map<String, String> rowMap = new HashMap<>();
-			Row row = rowIterator.next();
-			Iterator<Cell> cellIterator = row.cellIterator();
-			while (cellIterator.hasNext()) {
-				Cell cell = cellIterator.next();
-
-				int columnIndex = cell.getColumnIndex();
-				if (columnIndex < headers.size()) {
-					String header = headers.get(columnIndex);
-
-					if (!Strings.isNullOrEmpty(header)) {
-						// Need to ignore empty headers.
-						if (cell.getCellTypeEnum()
-								.equals(CellType.NUMERIC)) {
-								/*
-								This is a special handler for number cells.  It was requested that numbers
-								keep their formatting from their excel files.  E.g. 2.222222 with formatting
-								for 2 decimal places will be saved as 2.22.
-								 */
-							DataFormatter formatter = new DataFormatter();
-							String value = formatter.formatCellValue(cell);
-							rowMap.put(header, value);
-						} else {
-							cell.setCellType(CellType.STRING);
-							rowMap.put(header, cell.getStringCellValue());
+		return ImmutableMap.of("result", "complete");
+	}
+
+	/**
+	 * Save uploaded metadata to the
+	 *
+	 * @param locale    {@link Locale} of the current user.
+	 * @param session   {@link HttpSession}
+	 * @param projectId {@link Long} identifier for the current project
+	 * @return {@link Map} of potential errors.
+	 */
+	public Map<String, Object> saveProjectSampleMetadata(Locale locale, HttpSession session, long projectId) {
```

Comment:
Instead of returning a Map<String, Object> lets change this to use a custom dto class called `ProjectSampleMetadataUploadReponse` or something which has the following fields and getter/setter methods:

private String messageKey;
private String message;
private List<String> errorList;

Then this method would return a `ProjectSampleMetadataUploadReponse` which has the fields set below as required

Then in the controller call to this method you can replace with this:

	ProjectSampleMetadataReponse response = metadataImportService.saveProjectSampleMetadata(locale, session, projectId);

	if (response.getMessageKey.equals("success")) {
		return ResponseEntity.ok(response.getMessage());
	} else if(response.getMessageKey().equals("save-errors")) {
		return ResponseEntity.badRequest().body(response.getErrorList);
	} else {
		return ResponseEntity.badRequest()
				.body(response.getMessage());
	}

Reply:
Please see f96adb60601336c1db3004ece2a1e9c8f27a07a3.